### ISSUE TITLE
Add proper caching support

### ISSUE THREAD

**Issue by neithernut** [matthiasbeyer/imag#766]
In #765, the issue of search optimization was raised. I, once again, proposed to implement proper caching during that discussion to speed up searches. IMO, we should implement caching.
## Interface

Whether and how a cache is maintained should be configurable by the user, both in the configuration file and via the command-line. This mostly affects the (optional) creation of new cache entries for new store entries (or a subset of those entries). The cache should function as an index or dictionary, linking metadata to specific entries.

The cache should be used by search-/filter-functionality transparently. Also, a new `imag` subcommand `imag cache` should provide means to manage (clean, rebuild, drop, ...) the cache manually.
## Possible Implementation

The cache could be represented by a directory-structure:

The first levels in the directory would represent a hierarchy of metadata-fields. Below those, a file would be created for each value occurring. Those files would contain a list of entries having that metadata-filed set to that specific value, maybe in alphabetical order. Adding/removing/changing an entry would involve editing as many files as the entry has metadata-fields. To make better use of the file-system, a few intermediate directory-levels might be introduced to reduce the number of files in a directory. Also, the lists of entries might be compressed, since I/O is typically slower than (de)compressing, at least on most machines.

The filter would need a bit of attention. Using a cache we do _not_ have to iterate over _all_ the entries and filter them based on metadata. Instead, we can start from a relatively small set of entries, or multiple sets, and start filtering just those entries.

Some filters conceivable might be a bit tricky, e.g. when searching for mails with some substring in the title. But being able to start from an index containing those titles should still be faster than reading all the mails. Also, we can create an index on the metadata-values themselves. Maybe it would also make sense to introduce some query planner (like those used in SQL servers) in the long run.

Well, we could also just use a sqlite-database for caching. In this case we could just translate the filter expression (or at least a subset) to a SQL query and retrieve the list of entries.

In any case, search-/filter-interfaces should resort to filtering all the entries on-the-fly if no cache exists/is empty. That functionality _must_ be preserved. Maybe an option should be added instructing `imag` to resort to this procedure if the cache doesn't return any entries for a specific query, too.

Keeping the cache up-to-date could be done using hooks. This would also create some opportunities for users to tweak the behaviour.

---

I'm willing to implement this. However, I'm not sure how much time I will have during the next few months. On the other hand, I don't consider this an especially critical feature anyway, so it shouldn't matter that much if I need a bit longer.

Any suggestions or comments?

**Comment by matthiasbeyer** [comment]
> I'm willing to implement this. However, I'm not sure how much time I will have during the next few months. On the other hand, I don't consider this an especially critical feature anyway, so it shouldn't matter that much if I need a bit longer.

This is the key point. Everything I read sounds pretty okay to me, but as you say yourself, I do not consider this ciritical at all. Working with 1k maybe even 10k entries in the imag store shouldn't be much of a problem _with the current imag codebase_ even though it is not optimized for speet _at all_ (speaking of rather new storage devices as SSDs).
Working with 10k-100k entries could take some time, but only IO, so should still be doable, IMHO.

Caching would be relevant way past the 10k entries mark, IMO, so ... not critical at all, as you say.

---

Of course these numbers are based on my opinion, not on benchmarks. We clearly should do benchmarks before implementing caches or other optimizations. I would even say we should bench before we tweak the implementation for speed, as I think a nice-to-read codebase is way more important than speed (to a certain point, at least).

---

That beeing said, you're welcome to implement such things, we might have some more discussion, but I guess your work power (if you want to do something), could result in much more useful things at this point in time.

---

TL;DR: Ack if you want to play, if it works and is convenient to use I might merge it.

**Comment by neithernut** [comment]
Even with less than 10k of entries, caches might still be of interest in cases where searches are done frequently or in cases where metadata extraction is expensive. Also, considering possible future use-cases like managing photo-albums or (professional) media-libraries, one easily enters the realm of 100ks-of-entries.

Imagine a media-player which or an image-viewer featuring some search functionality. Waiting a few seconds for a query to yield a result might be OK. However, if you want to update the search results displayed to the user live, the search has to be completed within two seconds for the interface to "feel responsive". A few hundred milliseconds would be preferable. Its hard to say how much time a query takes without benchmarking, but I seriously doubt that the scan-and-filter-each-entry method would be fast enough in such a scenario.

Obviously, such a media-player doesn't exist yet --I planned to write one when you announced `imag`, but didn't yet have the time/motivation. However, sooner or later, *fast* lookups will be a feature of interest.

----

I'll probably start implementing a sqlite-based cache soonish. I'll try to come up with something this year.

**Comment by matthiasbeyer** [comment]
I'm closing this as I'm not expecting progress anymore.
